sc-kpm/sc-ui/translators/uiSc2SCnJsonTranslator.cpp

```diff
@@ -4,337 +4,633 @@
+        if (!(elInfoOutputArc->type & sc_type_arc_pos_const_perm && elInfoOutputArc->inputArcs.empty()))
+          continue;
+
+        structureElementIt = elInfoOutputArc->target->inputArcs.find(elInfoOutputArc);
+        if (structureElementIt != elInfoOutputArc->target->inputArcs.end())
+        {
+          // remove arc from source and target
+          elInfo->structureElements.insert(elInfoOutputArc->target);
+          elInfoOutputArc->target->inputArcs.erase(structureElementIt);
+          removableArcs.insert(elInfoOutputArc);
+        }
+      }
+      for (auto const & arc : removableArcs)
+      {
+        elInfo->outputArcs.erase(arc);
+      }
+
+      // get keyword from key elements
+      elInfo->structKeyword = findStructKeyword(keynodes);
+    }
+  }
 }
 
-void uiSCnSentenceNode::balance()
+sScElementInfo * uiSc2SCnJsonTranslator::resolveElementInfo(sc_addr const & addr, sc_type type)
 {
+  sScElementInfo * elInfo = mScElementsInfo[addr];
+  if (elInfo)
+    return elInfo;
+
+  if (type == 0)
+    sc_memory_get_element_type(s_default_ctx, addr, &type);
+
+  elInfo = new sScElementInfo();
+  elInfo->addr = addr;
+  elInfo->type = type;
+  elInfo->isInTree = false;
+  mScElementsInfo[addr] = elInfo;
+
+  return elInfo;
 }
 
-const String & uiSCnSentenceNode::json()
+sScElementInfo * uiSc2SCnJsonTranslator::findStructKeyword(sScElementInfo::tScElementInfoList const & structureElements)
 {
-  StringStream ss;
-
-  ss << "{";
-  ss << "\"id\": \"" << uiTranslateFromSc::buildId(mElementInfo->addr) << "\"";
-  ss << ", \"type\": " << mElementInfo->type;
+  sScElementInfo::tScElementInfoList structures;
+  sScElementInfo::tScElementInfoList elements = structureElements;
+  // first try to find structures in key elements
+  std::for_each(structureElements.cbegin(), structureElements.cend(), [&structures](sScElementInfo * el) {
+    if (el->type & sc_type_node_struct)
+      structures.insert(el);
+  });
+  if (!structures.empty())
+    elements = structures;
+
+  auto result = std::max_element(elements.begin(), elements.end(), [](sScElementInfo * left, sScElementInfo * right) {
+    return (left->outputArcs.size() + left->inputArcs.size()) < (right->outputArcs.size() + right->inputArcs.size());
+  });
+  if (result != elements.end())
+    return *result;
+
+  return nullptr;
+}
 
-  if (mType == ST_KEYWORD)
-    ss << ", \"keyword\": true";
+void uiSc2SCnJsonTranslator::json(sScElementInfo * elInfo, int level, bool isStruct, sc_json & result)
+{
+  bool isFullLinkedNodes = true;
 
-  if (mType == ST_PREDICATE)
+  getBaseInfo(elInfo, result);
+  if (!elInfo->structureElements.empty())
   {
-    assert(mChildSentences.size() == 1);
-    ss << ", \"SCNode\" : " << (*mChildSentences.begin())->json();
+    structureElements = elInfo->structureElements;
+    json(elInfo->structKeyword, 0, true, result[scnTranslatorConstants::STRUCT.data()]);
   }
-  else
+  // if node is arc
+  if (elInfo->type & sc_type_arc_mask)
```

Comment:
Let's for each structure element type create alone method. Decompose this (json) method on atomic methods